mitiq/zne/inference.py

```diff
@@ -1511,6 +1513,141 @@ def __eq__(self, other: Any) -> bool:
+        Calculates the expecation given a scale factor and model
+        parameters.
+
+        Args:
+            a: Model parameter.
+            b: Model parameter.
+            c: Model parameter.
+            scale_factors: The array of noise scale factors.
+
+        Returns:
+            The expected value according to the model parameters and
+            scale factor.
+        """
+
+        return a + b * np.exp(-1.0 * c * scale_factor)
+
+    @staticmethod
+    def extrapolate(
+        scale_factors: Sequence[float],
+        exp_values: Sequence[float],
+        full_output: bool = False,
+    ) -> Union[
+
+        float,
+    ]:
+        """Static method which evaluates an exponential extrapolation to the
+        zero-noise limit using Bayesian inference.
+
+        Args:
+            scale_factors: The array of noise scale factors.
+            exp_values: The array of expectation values.
+            full_output: If False (default), only the zero-noise limit is
+                returned. If True, additional information about the
+                extrapolated limit is returned too.
+        Returns:
+            zne_limit: The extrapolated zero-noise limit. If "full_output"
+                is False (default value), only this parameter is returned.
+            opt_params: The parameter array of the best fitting model.
+            zne_curve: The callable function which best fit the input data.
+                It maps a real noise scale factor to a real expectation value.
+                It is equal "zne_limit" when evaluated at zero.
+
+        Note:
+            This method computes the zero-noise limit only from the information
+            contained in the input arguments. To extrapolate from the internal
+            data of an instantiated Factory object, the bound method
+            ".reduce()" should be called instead.
+        """
+        with pm.Model():
+            """
+            We assumme that the priors for the model parameters is a uniform
+            distribution with an upper limit 1 and a lower limit 0,
+            except for b with a lower limit -1.
+            """
+            a = pm.Uniform('a', 0, 1)
+            b = pm.Uniform('b', -1, 1)
+            c = pm.Uniform('c', 0, 1)
+
+            # Noise parameter
+            eps = pm.Uniform('eps', 0, 0.5)
+
+            pm.Normal(
+                'expval',
+                mu=ExpBayesFactory._exp_ansatz(a, b, c, scale_factors),
+                sd=eps,
+                observed=exp_values,
+            )
+
+            trace = pm.sample(target_accept=0.95)
+
+        # Optimal model parameters:
+        a = trace['a'].mean()
+        b = trace['b'].mean()
+        c = trace['c'].mean()
+        opt_params, zne_error = [a, b, c], trace['eps'].mean()
+
+        def zne_curve(scale_factor: float) -> float:
+            return ExpBayesFactory._exp_ansatz(a, b, c, scale_factor)
+
+        zne_limit = ExpBayesFactory._exp_ansatz(a, b, c, 0.0)
```

Comment:
It looks like this uses the Bayesian estimate for the optimal parameters as inputs to the ansatz function in order to evaluate the extrapolated point, rather than reporting the Bayesian estimate of the extrapolated point itself. Especially with exponential functions, that can lead to a large systemic bias in estimates of the extrapolated value, since the function of an estimate isn't the same as the estimate of a function. That is, 𝔼ₓ[𝑓(𝑥)] ≠ 𝑓(𝔼ₓ[𝑥]) unless 𝑓 is an affine transformation.

https://iopscience.iop.org/article/10.1088/1367-2630/aaf207 has an example of that kind of an effect leading to a bias of about 10% in spectral density estimation:

![image](https://user-images.githubusercontent.com/31516/104131488-749bd180-532b-11eb-9c39-2bb0dc4624d1.png)

Reply:
Hi @cgranade, thanks for all the feedback! My first attempt was indeed to try and find the Bayesian estimate of the zero–noise limit. My approach was to use a hierarchical binomial model, but I found it difficult to make sense of how the model parameters and scale factors should be related to the conjugate prior probability distribution (Beta distribution). Do you have any suggestions as to how go about that approach, or maybe a completely new one? Also any relevant resources would be greatly appreciated.

Reply:
No worries, happy to help!

There's a number of different approaches that you can use, but effectively they all boil down to evaluating 𝔼[𝑎 + 𝑏 exp(−𝑐 𝑥₀) | data], where I'm using 𝑥₀ to represent the scale factor at which you're interested in extrapolating out to. For example, instead of substituting ̂𝑎 ≔ 𝔼[𝑎] into 𝑓(𝑥₀; 𝑎, 𝑏, 𝑐) = 𝑎 + 𝑏 exp(−𝑐 𝑥₀), you could sample the posterior that you computed above to get an (𝑎, 𝑏, 𝑐) triplet, substitute that into 𝑓(𝑥₀), and then treat that as a sample of the distribution over 𝑓(𝑥₀) conditioned on your data. Repeating that many times for many different samples for (𝑎, 𝑏, 𝑐) then gives an estimate for 𝑦₀ = 𝑓(𝑥₀).

As you mention, a hierarchical model can work as well. Effectively, you could set up the value 𝑦₀ = 𝑓(𝑥₀) as a new unknown parameter, using the delta-distribution Pr(𝑦₀ | 𝑎, 𝑏, 𝑐) = δ(𝑦₀ − 𝑓(𝑥₀; 𝑎, 𝑏, 𝑐)) to connect that new level of the hierarchy to the level that you've already defined here. If I understand correctly, PyMC3 represents that as the `Determinstic` class: https://docs.pymc.io/notebooks/api_quickstart.html#Deterministic-transforms Ideally, that should then let you sample directly from 𝑦₀, and to get both an estimate and uncertainty for your extrapolation using `trace['y0'].mean()` and by looking at the variance of `trace['y0']`.

I'll try to dig up some resources, but please feel free to ping again if there's anything else I can do in the meantime!